libkernelflinger/security_efi.c

```diff
@@ -123,62 +132,117 @@ BOOLEAN is_eom_and_secureboot_enabled(VOID)
 	return sbflags && enduser;
 }
 
-/* initially hardcoded all seeds as 0, and svn is expected as descending order */
-EFI_STATUS get_seeds(IN UINT32 *num_seeds, OUT VOID *seed_list)
+static EFI_STATUS bls_get_seed_list(BOOTLOADER_SEED_INFO_LIST *blist)
 {
 	EFI_STATUS ret = EFI_SUCCESS;
-	seed_info_t *tmp;
-	UINT32 i;
-#ifdef USE_TPM
-	UINT8 seed[TRUSTY_SEED_SIZE];
-#endif
 	BOOTLOADER_SEED_PROTOCOL *bls;
-	BOOTLOADER_SEED_INFO_LIST blist;
-
-	for (i = 0; i < BOOTLOADER_SEED_MAX_ENTRIES; i++) {
-		tmp = (seed_info_t *)(seed_list + i * sizeof(seed_info_t));
-		tmp->svn = BOOTLOADER_SEED_MAX_ENTRIES - i - 1;
-		memset(tmp->seed, 0, SECURITY_EFI_TRUSTY_SEED_LEN);
-	}
-	*num_seeds = BOOTLOADER_SEED_MAX_ENTRIES;
 
 	bls = get_bls_proto();
 	if (bls) {
-		ret = uefi_call_wrapper(bls->GetSeedInfoList, 1, &blist);
+		ret = uefi_call_wrapper(bls->GetSeedInfoList, 1, blist);
 		if (EFI_ERROR(ret)) {
 			efi_perror(ret, L"call GetSeedInfoList fail");
 			return ret;
 		}
 
 		debug(L"call GetSeedInfoList success");
-		*num_seeds = blist.NumOfSeeds;
-		ret = memcpy_s(seed_list, sizeof(blist.SeedList), blist.SeedList, sizeof(blist.SeedList));
-		memset(&blist, 0, sizeof(blist));
-		barrier();
+	}
+
+	return ret;
+}
+
+static UINT32 bls_get_max_svn_index(BOOTLOADER_SEED_INFO_LIST *blist)
+{
+        UINT32 i, max_svn_idx = 0;
+
+#ifdef USERDEBUG
+	/* if no seed found, use dummy seed for userdebug build */
+	if (blist->NumOfSeeds == 0) {
+		debug(L"No seed found, use dummy seed");
+		memset(blist, 0, sizeof(BOOTLOADER_SEED_INFO_LIST));
+		return 0;
+	}
+#endif
+        if ((blist->NumOfSeeds == 0) || (blist->NumOfSeeds > BOOTLOADER_SEED_MAX_ENTRIES))
+                return -1;
+
+        for (i = 1; i < blist->NumOfSeeds; i++) {
+                if (blist->SeedList[i].cse_svn > blist->SeedList[i - 1].cse_svn) {
```

Comment:
I am not sure whether following svn for SeedList is correct: SeedList[10] = { 1,2,5,6,3,4,}, in this condition, we will return max_svn_idx with 5 since SeedList[5] > SeedList[4]. However max svn is SeedList[3] = 6.
If above consumption is correct, then our logic is incorrect.
So, need confirmation from your side.

Reply:
nice catch